src/orange/surf/detail/InvoluteSolver.hh

```diff
@@ -0,0 +1,316 @@
+    CELER_FUNCTION real_type tmax() const { return tmax_; }
+
+  private:
+    //// DATA ////
+    // Involute parameters
+    real_type r_b_;
+    real_type a_;
+    real_type sign_;
+
+    // Bounds
+    real_type tmin_;
+    real_type tmax_;
+};
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from radius.
+ */
+CELER_FUNCTION InvoluteSolver::InvoluteSolver(
+    real_type r_b, real_type a, real_type sign, real_type tmin, real_type tmax)
+    : r_b_(r_b), a_(a), sign_(sign), tmin_(tmin), tmax_(tmax)
+{
+    double const pi = 3.14159265358979323846;
+    CELER_EXPECT(r_b > 0);
+    CELER_EXPECT(a > 0);
+    CELER_EXPECT(abs(tmax) < 2 * pi + abs(tmin));
+}
+
+//---------------------------------------------------------------------------//
+
+/*!
+ * Find all positive roots for involute surfaces that are within the bounds.
+ */
+CELER_FUNCTION auto
+InvoluteSolver::operator()(real_type x,
+                           real_type y,
+                           real_type z,
+                           real_type u,
+                           real_type v,
+                           real_type w) const -> Intersections
+{
+    double const pi = 3.14159265358979323846;
+
+    // Lambda used for calculating the roots
+    auto root = [](real_type t,
+                   real_type x,
+                   real_type y,
+                   real_type u,
+                   real_type v,
+                   real_type r_b_,
+                   real_type a_) {
+        real_type a = u * std::sin(t + a_) - v * std::cos(t + a_);
+        real_type b = t * (u * std::cos(t + a_) + v * std::sin(t + a_));
+        real_type c = r_b_ * (a - b);
+        return c + x * v - y * u;
+    };
+    /*
+     * Results initalization
+     */
+    Intersections result;
+    result = {no_intersection(), no_intersection(), no_intersection()};
+
+    if (u == 0 && v == 0)
+    {
+        return result;
+    }
+
+    Array<real_type, 3> dist;
+    real_type j = 0;
+
+    real_type convert = sqrt(ipow<2>(v) + ipow<2>(u) + ipow<2>(w))
+                        / sqrt(ipow<2>(v) + ipow<2>(u));
+
+    /*
+     * Define tolerances.
+     */
+    real_type const tolPoint = 1e-7;
+    real_type const tolConv = 1e-8;
+
+    // 0 distance of particle on invoute surface
+
+    real_type const rxy2 = ipow<2>(x) + ipow<2>(y);
+    real_type const tPoint = sqrt((rxy2 / (ipow<2>(r_b_))) - 1);
+    real_type angle = tPoint + a_;
+    real_type xInv = r_b_ * (std::cos(angle) + tPoint * std::sin(angle));
+    real_type yInv = r_b_ * (std::sin(angle) - tPoint * std::cos(angle));
+    if (abs(x - xInv) < tolPoint && abs(y - yInv) < tolPoint)
+    {
+        dist[j] = 0;
+        j++;
+    }
+
+    real_type beta;
+    // Line angle parameter
+    if (u != 0.0)
```

Comment:
Since we allow ORANGE/Celeritas to work with both double *and* single precision numbers, use `0` here to automatically promote the int to `real_type` rather than promote `u` to `double`.